pkg/runtime/mac/mac.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/elotl/itzo/pkg/metrics"
+	"strings"
+	"sync"
+)
+
+const (
+	vmImagePrefix = "mac-anka-img"
+)
+
+type MacRuntime struct {
+	metrics.AnkaMetricsProvider
+	registryClient RegistryClient
+	cliClient      *AnkaCLI
+	UnitsVMIDs     *sync.Map
+}
+
+func NewMacRuntime(registryClient RegistryClient) *MacRuntime {
+	executor := &CMDExecWrapper{}
+	return &MacRuntime{
+		registryClient: registryClient,
+		cliClient:      NewAnkaCLI(executor),
+		UnitsVMIDs:     new(sync.Map),
+	}
+}
+
+func (m *MacRuntime) RunPodSandbox(spec *api.PodSpec) error {
+	return m.cliClient.EnsureAnkaBin()
+}
+
+func (m *MacRuntime) StopPodSandbox(spec *api.PodSpec) error {
+	return nil
+}
+
+func (m *MacRuntime) RemovePodSandbox(spec *api.PodSpec) error {
+	return nil
+}
+
+func (m *MacRuntime) CreateContainer(unit api.Unit, spec *api.PodSpec, podName string, registryCredentials map[string]api.RegistryCredentials, useOverlayfs bool) (*api.UnitStatus, error) {
+	// check if image exists in registry
+	imageID, err := getVMImageIDFromUnitImage(unit)
+	if err != nil {
+		return api.MakeFailedUpdateStatus(unit.Name, unit.Image, "InvalidImgFormat"), err
+	}
+	err = m.registryClient.GetVMTemplate(imageID)
+	if err != nil {
+		return api.MakeFailedUpdateStatus(unit.Name, unit.Image, "VMTemplateNotFound"), err
+	}
+	err = m.cliClient.PullImage(imageID)
+	if err != nil {
+		return api.MakeFailedUpdateStatus(unit.Name, unit.Image, "VMTemplatePullFailed"), err
+	}
+
+	_, err = m.cliClient.Create(imageID)
+	if err != nil {
+		return api.MakeFailedUpdateStatus(unit.Name, unit.Image, "VMCreateFailed"), err
+	}
+	return api.MakeStillCreatingStatus(unit.Name, unit.Image, "VMTemplatePulled"), nil
+}
+
+func (m *MacRuntime) StartContainer(unit api.Unit, spec *api.PodSpec, podName string) (*api.UnitStatus, error) {
+	vmIDInterface, ok := m.UnitsVMIDs.Load(unit.Name)
+	if !ok {
+		return nil, fmt.Errorf("cannot find vm id for unit %s", unit.Name)
+	}
+	vmID := vmIDInterface.(string)
+	unitStatus, err := m.cliClient.Start(vmID)
+	if err != nil {
+		return nil, err
+	}
+	if unitStatus.Status != AnkaStatusOK {
+		return api.MakeFailedUpdateStatus(unit.Name, unit.Image, "VMStartFailed"), fmt.Errorf("cannot start vm: %s", unitStatus.Message)
+	}
+	started := true
+	return &api.UnitStatus{
+		Name: unit.Name,
+		State: api.UnitState{
+			Running: &api.UnitStateRunning{StartedAt: api.Now()},
+		},
+		Image:   unit.Image,
+		Ready:   true,
+		Started: &started,
+	}, nil
+}
+
+func getVMImageIDFromUnitImage(unit api.Unit) (string, error) {
+	// Let's set a convention that if container image has prefix mac-anka-img:
+	// then we'll treat suffix as anka VM Image ID.
+	// e.g. mac-anka-img:<vm-id>
+	vmImgName := strings.Split(unit.Image, ":")
+	if len(vmImgName) < 2 {
+		return "", fmt.Errorf("image has to be in format %s:vm-id", vmImagePrefix)
+	}
+	if vmImgName[0] != vmImagePrefix {
+		return "", fmt.Errorf("image has to be in format %s:vm-id", vmImagePrefix)
+	}
+	return vmImgName[1], nil
+}
```

Comment:
Do we really need this function? The assumption that we use a different image registry when the pod is supposed to run on Mac is implicit. If we used the anka registry URL + image path in `unit.Image` instead, we could do away from another command line parameter (registry URL), e.g. `localhost:8089/full/image/path`.

Reply:
hmm, image "name" would be always an uuid. If we would like to be really explicit, pod image could be set as:
`anka-registry.default.svc:8089/registry/v2/vm?id=234dsfasdfasdf`. However, from UX point of view, user need to know internals of anka registry (url path), which is less ideal. I like idea of skipping command line parameter. I'll try to strike a balance.

Reply:
So how does a the `image` field in units look now for anka?

Reply:
Also, the URL path part is fixed for a particular endpoint, so that does not need to be part of `unit.image`, but the registry base hostname should be. It's pretty simple to make sure we handle multiple registries here.

The same way as with regular linux containers: the docker-style image string is a simplification, and does not contain actual URL endpoint path details, but allows one to use any registry and image from that registry.